### PR TITLE
Major refactoring

### PR DESCRIPTION
**Pull request by anfelor**
This PR is a major refactoring of Parts.elm
It replaces all the functions with these nine:
- collection
  - create a Collection with encapsulates all Dict logic.
- accessors
  - create Accessors, like before
- create
  -  Construction of viewable compononts
  - render = create view find
- apply
  - Construction of message passing function
  - Unlike before, the components now have to specify a pass function
  - pass = apply update find
  - CounterMsg msg' -> Counter.pass CounterMsg msg' model
- apply'
  - `apply` for components, which `update` function returns the model as a Maybe,
    depending on whether it changed or not.
- accessors1, create1, apply1, apply1'
  - Like above, but for components with only one instance.

This slightly increases the code size of the components,
but more importantly it completely hides Dict from everybody and Parts from the component user.

Issue: Unfortunatly collections are way less useful than they should be due to [#520](https://github.com/elm-lang/elm-compiler/issues/520) 

EDIT: Also it has elm-lang/html as a dependency, which hopefully shouldn't be a issue as this library is so view centric.

### CODE REVIEW

**Comment by debois** [comment]
Hi @anfelor, 

This is great! The elm-parts library has been needing a refactor for some time now, but I couldn't find a good way to do it. I really like your approach.

As far as I can tell, your PR forces the client to have messages and update clauses for each type of Part (I've made a line note at the point where I think I can see this). The key goal of elm-parts is to reduce boilerplate enough that a UI library (like [elm-mdl](https://github.com/debois/elm-mdl)) can be as convenient to use as core/html.  

But in the PR users would not be able to get away with having just a single boilerplate UI-library message and clause, but need instead one for every component they use (Button, Toggle, Menu, Layout, ...). That is a regression on the as-convenient-as-core/html goal. I'd _really_ like it if we could find a way to avoid this regression?

(I made a few other comments as line notes.)

File: examples/0-counter-part.elm [hunk 1]
```diff
@@ -21,13 +19,13 @@ main =
...
 type alias Model =
-  { counter : Indexed Counter.Model 
+  { counters : Counters
   }
 
 
 init : (Model, Cmd Msg)
 init =
-  ( { counter = Dict.empty }
+  ( { counters = .empty Counters.all }
```
Comment (debois) [inline examples/0-counter-part.elm]: Neat.

File: examples/0-counter-part.elm [hunk 2]
```diff
@@ -47,7 +45,7 @@ update msg model =
       init
 
     CounterMsg msg' -> 
-      Parts.update CounterMsg msg' model
+      Counters.pass CounterMsg msg' model
```
Comment (debois) [inline examples/0-counter-part.elm]: This is the part where I'm thinking that every new component will require the user to put in an extra clause in `update`: Suppose I also have a `Button` component. I'd need a boilerplate clause

```
ButtonMsg msg' -> 
  Buttons.pass ButtonMsg msg' model 
```

Right? For UI libraries, I might be using Buttons, Checkbox, Toggles, Tooltips, Tabs, Sliders... Lots of boilerplate.

File: examples/Counters.elm [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+render 
+   : ((Msg, Index) -> outerMsg) 
+  -> Index
+  -> Container c
+  -> Html outerMsg
+render =
+  create view find
+
+all : Collection Model (Container c)
+all =
```
Comment (debois) [inline examples/Counters.elm:92]: It's great that `all` is a top-level value (as opposed to a function). One thing that I never liked about Parts is that all the mechanics for creating getters and setters and embedded update functions is repeated on every call to `render`.

File: src/Parts.elm [hunk 4]
```diff
@@ -87,268 +135,222 @@ type alias Update' model msg =
...
-  ( \idx c -> Dict.get idx (get c) |> Maybe.withDefault model0
-  , \idx model c -> set (Dict.insert idx model (get c)) c
-  )
-  
-
--- EMBEDDING MESSAGES
-
+ -> Index
+ -> Accessors model container
+accessors collection model0 idx =
```
Comment (debois) [inline src/Parts.elm:196]: It'd be neat if `accessors collection model0` had already computed `empty`, `get`, `set` etc. so that when you add the index, as little further computation as necessary is performed. Don't know if that's possible.

File: src/Parts.elm [hunk 5]
```diff
@@ -74,7 +122,7 @@ type alias Update model msg =
 {-| Standard TEA view function type. 
 -}
 type alias View model a = 
-  model -> a
+  model -> Html a
```
Comment (debois) [inline src/Parts.elm:125]: I think this is overly restrictive. View functions need arguments (e.g., I'm making a button component and I want to have a `disabled` option).

File: src/Parts.elm [hunk 6]
```diff
@@ -87,268 +135,222 @@ type alias Update' model msg =
...
-      |> map2nd (Cmd.map (pack' upd)))
+create
+   : View model innerMsg 
+  -> (Index -> Accessors model container)
+  -> ((innerMsg, Index) -> outerMsg)
+  -> Index
+  -> container
+  -> Html outerMsg
+create view access wrapper idx container = 
+  App.map (\msg -> wrapper (msg, idx)) (view ((access idx).get container))
```
Comment (debois) [inline src/Parts.elm:251]: I see why you restricted the type `View`. Here's an idea: Push the `App.map` into the `view` function (i.e., we require of `view` that it knows itself how to lift its messages, perhaps by applying `App.map` internally). The type of View is now

```
type alias View = (Msg -> m) -> Model -> a
```

Then `create` should look something like

```
create view access wrapper idx container = 
  view (\msg -> wrapper (msg, idx)) ((access idx).get) container
```

Obviously, this means we now expect `view` functions to have non-standard shape; this is not too bad since we're using a known trick (this was up on elm-discuss).

File: examples/0-counter-part.elm (hunk 1 above)
Comment (anfelor) [inline examples/0-counter-part.elm]: Unfortunatly, this yields a very obscure error message, when writing `Counters.all.empty`, see [#520](https://github.com/elm-lang/elm-compiler/issues/520)

File: examples/0-counter-part.elm (hunk 2 above)
Comment (anfelor) [inline examples/0-counter-part.elm]: Isn't this already the case in elm-mdl if you don't want all components? http://package.elm-lang.org/packages/debois/elm-mdl/6.1.0/Material#optimising-for-size

I'll add a example showing, how to "unify" these.

File: src/Parts.elm (hunk 4 above)
Comment (anfelor) [inline src/Parts.elm:196]: Will look into it.

File: src/Parts.elm (hunk 5 above)
Comment (anfelor) [inline src/Parts.elm:125]: Shouldn't that be in the model? I am relatively new to elm, so an example would be helpful.

File: examples/0-counter-part.elm (hunk 2 above)
Comment (anfelor) [inline examples/0-counter-part.elm]: Ah, I see the problem (and difference to the existing solution) now.

I already have an idea of how to fix it, but it involves even more functions. I'll think about it.

**Comment by anfelor** [comment]
I added an example of a "unifying" component. But I don't like that solution and will keep investigating.

File: src/Parts.elm (hunk 5 above)
Comment (debois) [inline src/Parts.elm:125]: No, I think it's important to make the distinction between "parameters to the view" and "parameters to the model". The main user of elm-parts, [debois/elm-mdl](https://github.com/debois/elm-mdl) pushes as much customisation as possible onto view rather than into the model. See, e.g., the "getting started" example [here](https://github.com/debois/elm-mdl/blob/master/examples/Counter.elm). (`render` is the output of `create`.)

**Comment by debois** [comment]
Ok. Could you say a little bit about what you updated, where I should look, and why you don't like it?

**Comment by debois** [comment]
Hi @anfelor, I had to update elm-parts to accommodate elm-mdl. In the new version, update must take a lifting function, and we now require it to do explicit no-ops:

``` elm
update : (Msg -> msg) -> Msg -> Model -> Maybe (Model, Cmd msg)
```

(There is a convenience function for lifting a standard `update` to this form.)

This change was necessary to allow the `update` function to _either_ treat its own commands using the first argument lifting function, _or_ dispatch commands for its parent, which it has gotten through either model or Msg. E.g., 

``` elm
update fwd msg model = 
  case msg of 
     MyInternalMsg -> 
        ...
        ( model', Cmd.map fwd myInternalCmd ) -- myInternalCmd : Cmd Msg

     MyMessageWhichShouldSendACommandOfParent (parentCmd) ->
       ...
       ( model', parentCmd ) -- parentCmd : Cmd msg
```

I have the vague idea that this change may make it easier to apply your refactoring. Could you take a look?

**Comment by anfelor** [comment]
Well, I'll be experimenting a bit. My latest idea was https://github.com/anfelor/elm-parts/commit/acb3b8a16c6ecc5729d1e9de2420bb7768b9fb6a#diff-8d0b161e39fc2c00dc09fc4b51cfc9cb, but (once again) I don't like that solution.

To my two commits: Nothing new in Parts.elm, I simply added an extra layer [Controls.elm](https://github.com/anfelor/elm-parts/blob/cebbc4d0d37049c234350bbb66ce120c2aa38a9e/examples/Controls.elm) to the examples, that bundles the two components Counter and Help. The Messages are leaking through, however: 

``` elm
Counters.render (ControlMsg << Controls.CounterMsg) [0] model.controls
```

This has motivated the commit above on the [objects branch](https://github.com/anfelor/elm-parts/tree/objects). It turns components into objects, that could be modified by the control layer, but I have yet to implement that part.

Currently I am thinking, that TEA should maybe be more event based.. but once I have something working, I will inform you!

**Comment by debois** [comment]
Ok, thanks. 

I'm not sure exactly where you are going, but just in case I should probably say this: if your refactoring moves too far away from (a) TEA and (b) the needs of elm-mdl, I _might_ be reluctant to merge it. 

It looks exciting and interesting, though. I'd be really interested in seeing what you come up with.

**Comment by anfelor** [comment]
As you probably have already guessed I stopped working on this some time ago. Mainly I wanted to create a way of creating sub-components without subscriptions and commands; components that hide their inner workings, while at the same time remaining easy to test. Actually I like to think of HTML that way: what is a button if not such an component?

But in order to do that I would have needed to replace elm-html, while still using only elm-virtualdom to get that library into the official package manager. I would even be willing to do that (unpaid) work, but I feel increasingly less at home in the elm community. It seems to me that Evan and NoRedInk have a monopoly on the design of elm apps. This is a good thing on the one hand: Javascripters want additions that are unnecessary in functional programming, haskellers advanced type system features, that no Javascripter has the time to understand.

On the other hand, they are claiming to be the only ones having a clue, because they have access to thousends of lines of NoRedInk's Elm code. I agree that Elm can be easier to read than Haskell.  But I also know, that Haskell is a lot more expressive and that can lead to better code too! With Haskell I have a choice with Elm I haven't (of course this is limited by the libraries and no one can use Haskell without knowledge of monads. But Elm is full of monads too, they only call them 'good api design').

And this leads to the current situation, where lots of issues go unresolved because they don't want to make a mistake. They don't want to add type classes, they don't want to add components and most importantly they don't want to give the programmers the freedom to add them themselves. This however means for me, that Elm will never become a bigger programming language: Who would be using JS now if it weren't for (imperfect) community libraries like jQuery or Angular?  You might argue now, that Elm has changed considerably between 0.16 to 0.17, while vanilla JS didn't change at all. But what happens once Elm reaches 1.0?

To me a language should provide a stable ecosystem to help the developers write their libraries. Elm however is more a framework then a language and therefore extremely restricting. I believe the fundamental strength of functional programming doesn't lie in the simplicity of immutability, but in the absence of many predefined constructs. Java programmers had to wait for years to get the for each loop construct, a problem that only exist in the more experimental areas of Haskell, when you depend on GHC extensions. Elm however introduces many predefined functions like any framework and so rules out any contribution from the community, that doesn't depend on tricks.

With this letter I am therefore also declaring my leave from the Elm community. Elm has shown me, that replacing `\x -> f x x` with `join f` doesn't improve code after all and also what I truly value about functional programming. Elm has a great community, but I think I have better tools to work with.

**tldr;**
I am quitting this. No need to worry.
